pymuonsuite/io/castep.py

```diff
@@ -7,109 +7,348 @@
+        if calc_type == "GEOM_OPT" or calc_type == "MAGRES":
+            if sname is None:
+                sname = os.path.split(folder)[-1]  # Same as folder name
+
+            self._calc = deepcopy(self._calc)
+
+            # We only use the calculator attached to the atoms object if a calc
+            # has not been set when initialising the ReadWrite object OR we
+            # have not called write() and made a calculator before.
+
+            if self._calc is None:
+                if isinstance(a.calc, Castep):
+                    self._calc = deepcopy(a.calc)
+                self._create_calculator()
+
+            self._update_calculator(calc_type)
+            a.set_calculator(self._calc)
+
+            io.write(os.path.join(folder, sname + '.cell'),
+                     a, magnetic_moments='initial')
+            write_param(os.path.join(folder, sname + '.param'),
+                        a.calc.param, force_write=True)
+
+            if self.script is not None:
+                stxt = open(self.script).read()
+                stxt = stxt.format(seedname=sname)
+                with open(os.path.join(folder, 'script.sh'), 'w') as sf:
+                    sf.write(stxt)
+        else:
+            raise(NotImplementedError("Calculation type {} is not implemented."
+                  " Please choose 'GEOM_OPT' or 'MAGRES'".format(calc_type)))
+
+    def _create_calculator(self):
+        if self._calc is not None and isinstance(self._calc, Castep):
+            calc = deepcopy(self._calc)
+        else:
+            calc = Castep()
+
+        mu_symbol = self.params.get('mu_symbol', 'H:mu')
+
+        # Start by ensuring that the muon mass and gyromagnetic ratios are
+        # included
+        gamma_block = calc.cell.species_gamma.value
+        if gamma_block is None:
+            calc.cell.species_gamma = add_to_castep_block(gamma_block,
+                                                          mu_symbol,
+                                                          constants.m_gamma,
+                                                          'gamma')
+
+            mass_block = calc.cell.species_mass.value
+            calc.cell.species_mass = add_to_castep_block(mass_block, mu_symbol,
+                                                         constants.m_mu_amu,
+                                                         'mass')
+
+        # Now assign the k-points
+        calc.cell.kpoint_mp_grid = list_to_string(
+            self.params.get('k_points_grid', [1, 1, 1]))
+
+        # Read the parameters
+        pfile = self.params.get('castep_param', None)
+        if pfile is not None:
+            calc.param = read_param(self.params['castep_param']).param
+
+        self._calc = calc
+
+        return self._calc
+
+    def _update_calculator(self, calc_type):
+        if calc_type == "MAGRES":
+            # check if our calculator is already set up for Magres
+            # if it is, we don't need to modify the calc.
```

Comment:
This is a very good point!

I believe changing line 272 to this:

`if not (self._calc.param.task == 'Magres' and self._calc.param.magres_task = 'Hyperfine'):`

would resolve this - I don't think there are any other settings we have to add. Honestly I am not sure if there are other settings that by having them would cause a problem. i.e. if in their calculator they had set it to magres, hyperfine, and then some other random settings we wouldn't usually set. But this could happen anyway as we always read in their parameter file and keep what's been set in there other than what we specifically need set...

Reply:
See: https://github.com/muon-spectroscopy-computational-project/pymuon-suite/pull/7#discussion_r570187129